File: basic_3/TwoSum.java
import java.util.*;
public class TwoSum {
    public static int[] twoSum(int[] arr, int target) {

        for (int i = 0; i < arr.length; i++) {
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[i] + arr[j] == target) {
                    return new int[] { i, j };
                }
            }
        }
        return new int[] { -1, -1 };
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter array size: ");
        int n = sc.nextInt();

        int[] arr = new int[n];

        System.out.println("Enter array elements:");

        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        System.out.print("Enter target: ");
        int target = sc.nextInt();

        int[] result = twoSum(arr, target);

        if (result[0] == -1)
            System.out.println("No Pair Found");
        else
            System.out.println("Indices: " + result[0] + " " + result[1]);

        sc.close();
    }
}

File: DSA/TwoSum.py
n = int(input())

arr = list(map(int, input().split()))

q = int(input())

for _ in range(q):

    target = int(input())

    seen = set()
    found = False

    for num in arr:

        required = target - num

        if required in seen:
            found = True
            break

        seen.add(num)

    if found:
        print("YES")
    else:
        print("NO")


"""
Input
5
2 7 11 15 2
4
9
20
4
100
Output
YES
NO
YES
NO
     """



# n = int(input())

# arr = list(map(int, input().split()))

# q = int(input())

# for _ in range(q):

#     target = int(input())

#     frequency = {}

#     count = 0

#     for num in arr:

#         required = target - num

#         count += frequency.get(required, 0)

#         frequency[num] = frequency.get(num, 0) + 1

#     print(count)


"""
Input:

4
2 2 2 7
1
4

Output:

3

This version is NOT asking whether a pair exists.

It is asking HOW MANY pairs exist.
"""




n = int(input())

arr = list(map(int, input().split()))

target = int(input())

seen = {}

found = False

for i in range(n):

    required = target - arr[i]

    if required in seen:
        print(seen[required], i)
        found = True
        break

    if arr[i] not in seen:
        seen[arr[i]] = i

if not found:
    print(-1, -1)


"""
The question says:

Find the indices of two distinct elements whose sum equals the target.

Example:

Array = [2, 7, 11, 15]

Target = 9

Answer:

0 1

Because:

arr[0] + arr[1]
= 2 + 7
= 9
 """